### PR TITLE
Implement UsortVerifyMultiplicityBody hint

### PR DESCRIPTION
**Pull request by wugalde19**
Resolves https://github.com/NethermindEth/cairo-vm-go/issues/235

### CODE REVIEW

File: pkg/hintrunner/zero/zerohint_usort_test.go [hunk 1]
```diff
@@ -21,5 +22,79 @@ func TestZeroHintUsort(t *testing.T) {
 				check: varValueInScopeEquals("__usort_max_size", feltUint64(1)),
 			},
 		},
+		"UsortVerifyMultiplicityBody": {
+			{
+				ctxInit: func(ctx *hinter.HintRunnerContext) {
+					// Having empty scopes in scope manager will cause an error
```
Comment (MaksymMalicki) [inline pkg/hintrunner/zero/zerohint_usort_test.go:28]: There is no need to test here if the scope manager is empty, as the function is always invoked in the scope of usort() function

File: pkg/hintrunner/zero/zerohint_usort_test.go [hunk 2]
```diff
@@ -21,5 +22,79 @@ func TestZeroHintUsort(t *testing.T) {
...
+				},
+				ctxInit: func(ctx *hinter.HintRunnerContext) {
+					ctx.ScopeManager.EnterScope(map[string]any{
+						"positions": []uint64{1, 2, 3},
+					})
+				},
+				makeHinter: func(ctx *hintTestContext) hinter.Hinter {
+					return newUsortVerifyMultiplicityBodyHint(ctx.operanders["next_item_index"])
+				},
+				check: varValueInScopeEquals("current_pos", feltUint64(3)),
```
Comment (MaksymMalicki) [inline pkg/hintrunner/zero/zerohint_usort_test.go:66]: Maybe also add check if the last_pos value is as expected in scope

Comment (MaksymMalicki) [reply pkg/hintrunner/zero/zerohint_usort_test.go:66]: You can then remove the next test. There are utilities like ```allVarValueInScopeEquals``` for that

File: pkg/hintrunner/zero/zerohint_usort.go [hunk 3]
```diff
@@ -26,3 +30,63 @@ func newUsortEnterScopeHint() hinter.Hinter {
...
+				return fmt.Errorf("getting positions from scope failed: %w", err)
+			}
+
+			positions, ok := positionsInterface.([]uint64)
+			if !ok {
+				return fmt.Errorf("casting positions into an array of uint64 failed: %w", err)
+			}
+
+			current_pos := utils.Pop(&positions)
+			last_pos := current_pos + 1
```
[the remark this answers is not in the document]
Comment (wugalde19) [reply pkg/hintrunner/zero/zerohint_usort.go:53]: You're right
I was confused because it didn't make sense to me that we were using a variable that was not declared but now that I understand more the code I noticed that `last_pos` is present in the scope and we just need to get the value from there

Thanks for pointing this out @MaksymMalicki

File: pkg/hintrunner/zero/zerohint_usort_test.go [hunk 4]
```diff
@@ -21,5 +23,64 @@ func TestZeroHintUsort(t *testing.T) {
 				check: varValueInScopeEquals("__usort_max_size", feltUint64(1)),
 			},
 		},
+		"UsortVerifyMultiplicityBody": {
+			{
```
Comment (MaksymMalicki) [inline pkg/hintrunner/zero/zerohint_usort_test.go:27]: you can add some very short comments on what this test tests

File: pkg/hintrunner/zero/zerohint_usort_test.go [hunk 5]
```diff
@@ -21,5 +23,64 @@ func TestZeroHintUsort(t *testing.T) {
...
+				ctxInit: func(ctx *hinter.HintRunnerContext) {
+					ctx.ScopeManager.EnterScope(map[string]any{
+						"positions": []uint64{6, 4, 2},
+						"last_pos":  0,
+					})
+				},
+				makeHinter: func(ctx *hintTestContext) hinter.Hinter {
+					return newUsortVerifyMultiplicityBodyHint(ctx.operanders["next_item_index"])
+				},
+				check: allVarValueInScopeEquals(map[string]any{
```
Comment (MaksymMalicki) [inline pkg/hintrunner/zero/zerohint_usort_test.go:62]: I think you can merge this test and the bottom one into one. You can call two check functions in one test, by doing:
```
check: func(t *testing.T, ctx *hintTestContext) {					varAddrResolvedValueEquals("output_length", feltUint64(3))(t, ctx)
	allVarValueInScopeEquals()(t, ctx)
	allVarValueEquals()(t, ctx)
},
```

File: pkg/hintrunner/zero/zerohint_usort_test.go [hunk 6]
```diff
@@ -21,5 +23,64 @@ func TestZeroHintUsort(t *testing.T) {
...
+				ctxInit: func(ctx *hinter.HintRunnerContext) {
+					ctx.ScopeManager.EnterScope(map[string]any{
+						"positions": []uint64{6, 4, 2},
+						"last_pos":  1,
+					})
+				},
+				makeHinter: func(ctx *hintTestContext) hinter.Hinter {
+					return newUsortVerifyMultiplicityBodyHint(ctx.operanders["next_item_index"])
+				},
+				check: allVarValueEquals(map[string]*fp.Element{
```
Comment (MaksymMalicki) [inline pkg/hintrunner/zero/zerohint_usort_test.go:80]: when you merge the tests, you can use here ```varValueEquals``` instead of ```allVarValueEquals```

File: pkg/utils/arrays.go [hunk 7]
```diff
@@ -1,7 +1,22 @@
...
+
 func Reverse[T any](a []T) {
 	for i, j := 0, len(a)-1; i < j; i, j = i+1, j-1 {
 		a[i], a[j] = a[j], a[i]
 	}
 }
+
+func Pop[T any](a *[]T) (T, error) {
+	if len(*a) == 0 {
+		var zeroValue T
```
[the remark this answers is not in the document]
Comment (wugalde19) [reply pkg/utils/arrays.go:15]: @MaksymMalicki I had it like 
image

But it is giving me a warning
If you want I can just ignore it
image

I was also considering make the return type a pointer `*T` but not sure if that's the best approach

What do you think?

File: pkg/hintrunner/zero/zerohint_usort_test.go [hunk 8]
```diff
@@ -21,5 +22,54 @@ func TestZeroHintUsort(t *testing.T) {
...
+				},
+				errCheck: func(t *testing.T, ctx *hintTestContext, err error) {
+					require.NotNil(t, err)
+				},
+			},
+			// Tests when the positions interface cannot be casted to an array of uint64.
+			{
+				ctxInit: func(ctx *hinter.HintRunnerContext) {
+					ctx.ScopeManager.EnterScope(map[string]any{
+						"positions": []string{"6", "4", "2"},
```
Comment (har777) [inline pkg/hintrunner/zero/zerohint_usort_test.go:39]: the positions seem to be written as:
```
positions_dict = defaultdict(list)
for i in range(input_len):
    val = memory[input_ptr + i]
    positions_dict[val].append(i)
```

So no need to test them with strings. uint64 should suffice.

File: pkg/utils/arrays.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (MaksymMalicki) [reply pkg/utils/arrays.go:15]: No no, its okey, I was just wondering haha 😛. Returning pointer is not a good idea, as it will cause heap allocations and performance issues. The current solution looks great

File: pkg/hintrunner/zero/zerohint_usort.go [hunk 9]
```diff
@@ -26,3 +28,66 @@ func newUsortEnterScopeHint() hinter.Hinter {
...
+				return err
+			}
+
+			err = vm.Memory.WriteToAddress(&addrNextItemIndex, &newNextItemIndexMemoryValue)
+			if err != nil {
+				return err
+			}
+
+			// Save `current_pos` and `last_pos` values in scope variables
+			err = ctx.ScopeManager.AssignVariables(map[string]any{
```
Comment (MaksymMalicki) [inline pkg/hintrunner/zero/zerohint_usort.go:76]: Sorry, in previous comment I pointed that you can just return err. Actually you can just return what this function returns haha 😅

File: pkg/hintrunner/zero/zerohint_usort.go [hunk 10]
```diff
@@ -26,3 +28,66 @@ func newUsortEnterScopeHint() hinter.Hinter {
...
+			if !ok {
+				return err
+			}
+
+			current_pos, err := utils.Pop(&positions)
+			if err != nil {
+				return err
+			}
+
+			last_pos, err := ctx.ScopeManager.GetVariableValue("last_pos")
```
Comment (har777) [inline pkg/hintrunner/zero/zerohint_usort.go:55]: type case to `.(uint64)`

File: pkg/utils/arrays.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (wugalde19) [reply pkg/utils/arrays.go:15]: @MaksymMalicki ok cool
Thanks for the explanation

**Comment by MaksymMalicki** [comment]
lgtm

**Comment by MaksymMalicki** [comment]
For some reason the tests are not passing, make sure to run ```make unit``` and resolve the problems. Thanks!

**Comment by wugalde19** [comment]
> For some reason the tests are not passing, make sure to run `make unit` and resolve the problems. Thanks!

@MaksymMalicki My apologies
Something was wrong with the environment
I dockerized it to make sure it does not happen again
image

I did fix the test and made a small change reading the `current_pos` from the scope instead of using the one from `positions` to calculate the value of `next_item_index` and `last_pos`

**Comment by wugalde19** [comment]
@MaksymMalicki I just fixed conflicts after rebasing with main branch

File: pkg/hintrunner/zero/zerohint_usort.go [hunk 11]
```diff
@@ -107,3 +108,74 @@ func createUsortVerifyHinter(resolver hintReferenceResolver) (hinter.Hinter, err
...
+			if err != nil {
+				return err
+			}
+
+			positions, ok := positionsInterface.([]int)
+			if !ok {
+				return fmt.Errorf("cannot cast positionsInterface to []int")
+			}
+
+			new_current_pos, err := utils.Pop(&positions)
```
Comment (cicr99) [inline pkg/hintrunner/zero/zerohint_usort.go:130]: Could you use camelCase for variable names on the go side?

File: pkg/hintrunner/zero/zerohint_usort.go [hunk 12]
```diff
@@ -107,3 +108,74 @@ func createUsortVerifyHinter(resolver hintReferenceResolver) (hinter.Hinter, err
...
+			//> current_pos = positions.pop()
+			//> ids.next_item_index = current_pos - last_pos
+			//> last_pos = current_pos + 1
+
+			positionsInterface, err := ctx.ScopeManager.GetVariableValue("positions")
+			if err != nil {
+				return err
+			}
+
+			positions, ok := positionsInterface.([]int)
```
Comment (cicr99) [inline pkg/hintrunner/zero/zerohint_usort.go:125]: It's better to specify int64 for this

File: pkg/hintrunner/zero/zerohint_usort.go (hunk 10 above)
Comment (cicr99) [reply pkg/hintrunner/zero/zerohint_usort.go:145]: the idea is to do the typecasting here for `last_pos` the same way you did for `current_pos`

**Review by cicr99 — COMMENTED** [review]
Just these minor details left. The PR looks great!

File: pkg/hintrunner/zero/zerohint_usort.go (hunk 12 above)
Comment (wugalde19) [reply pkg/hintrunner/zero/zerohint_usort.go:125]: @cicr99 I tried but it was throwing an error, so I had to use `uint64` instead

image

then for `currentPos` and `lastPos` values, I was also getting an error when trying to cast to `int64` so I keep them as `int`

image

Comment (wugalde19) [reply pkg/hintrunner/zero/zerohint_usort.go]: Understood and done
